Assets/Src/Shaders/GridShader.shader

```diff
@@ -0,0 +1,180 @@
+        float4 pos : POSITION;
+        float3 normal : NORMAL;
+        float2 uv : TEXCOORD0;
+      };
+
+      struct v2f {
+        float4 pos : SV_POSITION;
+        float3 worldNormal : TEXCOORD0;
+        float3 worldPos : TEXCOORD1;
+        float2 uv : TEXCOORD2;
+      };
+
+      sampler2D _MainTex;
+      float4 _MainTex_ST;
+      fixed4 _Color;
+      fixed4 _Specular;
+      float _Gloss;
+      fixed4 _GridColor;
+      float _GridLineWidth;
+      float _GridSize;
+      fixed4 _SecondaryGridColor;
+      float _SecondaryGridLineWidth;
+      float _SecondaryGridSize;
+
+      v2f vert(a2v v) {
+        v2f o;
+        o.pos = UnityObjectToClipPos(v.pos);
+        o.worldNormal = UnityObjectToWorldNormal(v.normal);
+        o.worldPos = mul(unity_ObjectToWorld, v.pos).xyz;
+        o.uv = TRANSFORM_TEX(v.uv, _MainTex);
+        return o;
+      }
+
+      // Calculates the grid color at a specified uv position.
+      fixed4 GetGridColor(
+          fixed4 gridColor,
+          float gridSize,
+          float gridLineWidth,
+          float2 uv) {
+        const float Pi = 3.14159;
+        const float GridLineDensity = 3000.0;
+        // Scales a periodic curve (y = cos(x)) to the expected period and amplitude, then clips an
+        // arc out of each period of the curve.
+        float density = GridLineDensity / gridSize;
+        float2 curve = abs(cos(gridSize * Pi * (uv - .5)));
+        float2 clip = saturate((curve - 1) * density / gridLineWidth + 1);
+        float blend = saturate(clip.x + clip.y);
+        fixed3 color = gridColor.rgb * blend;
+        // The alpha channel is not affected by the grid calculation.
+        return fixed4(color, gridColor.a);
+      }
+
+      // Returns the max number among the three components of a vector.
+      float Max3(fixed3 v) {
+        return max(max(v.x, v.y), v.z);
+      }
+
+      fixed4 frag(v2f i) : SV_Target {
+        // Texture and lighting calculation for a specular surface.
+        fixed3 worldNormal = normalize(i.worldNormal);
+        fixed3 worldLightDir = normalize(UnityWorldSpaceLightDir(i.worldPos));
+        fixed4 texColor = tex2D(_MainTex, i.uv);
+        fixed3 albedo = texColor.rgb * _Color.rgb;
+        fixed3 ambient = UNITY_LIGHTMODEL_AMBIENT.xyz * albedo;
+        fixed3 diffuse = _LightColor0.rgb * albedo * saturate(dot(worldNormal, worldLightDir));
+        fixed3 viewDir = normalize(UnityWorldSpaceViewDir(i.worldPos));
+        fixed3 halfDir = normalize(worldLightDir + viewDir);
+        fixed3 specular =
+            _LightColor0.rgb * _Specular.rgb * pow(saturate(dot(worldNormal, halfDir)), _Gloss);
+        fixed3 main = ambient + diffuse + specular;
+
+        // Prepares a grid color and a grayscale mask. The value of the mask indicates how much the
+        // grid color will be blended to the main color.
+        fixed4 grid = GetGridColor(_GridColor, _GridSize, _GridLineWidth, i.uv);
+        fixed gridMask = grid.a * Max3(grid);
```

Comment:
The mask is calculated by the alpha and the color value. If I set the grid color to black, then the line will be disappeared. This should not be the intended design.

How about adjust the alpha value of the line by multiplying the alpha of grid color and the curve value in the GetGridColor() function to make the edge of the line transparent?

Reply:
Oh, my usage of masks was not correct. Your suggestion is great! Multiplying the alpha value and the curve value generates the intended result.

Updated the code logic and please have another look.